### ISSUE TITLE
Incorrect gradient when arguments to minimum() and maximum() are equal

### ISSUE THREAD

**Issue by astooke** [Theano/Theano#6470]
Hi,
 
EDITED: previously had written about `theano.tensor.clip()` in here, but that was not the problem.

I'm experiencing incorrect gradients when the loss function includes `theano.tensor.minimum()` and the two arguments to `minimum` are equal.  It returns either the sum, or one of the values multiplied by 2.  

See the example below.  Comparing all the pieces against the same in Tensorflow, only the last assertion--the final gradient--fails for me when `np_ratios1` is used for both inputs.

It also happens in `theano.tensor.maximum()`.

Please Help!!!

Happens in Theano 0.9 and bleeding edge 0.10.

```
import numpy as np
import theano
import theano.tensor as T
import tensorflow as tf

SIZE = 10

# NUMPY

np_ratios1 = 0.4 * np.random.randn(SIZE).astype('float32') + 1
np_ratios1 = np.clip(np_ratios1, 0.1, 100)
np_ratios2 = 0.4 * np.random.randn(SIZE).astype('float32') + 1
np_ratios2 = np.clip(np_ratios2, 0.1, 100)
np_advs = np.random.randn(SIZE).astype('float32')
np_surr1 = np_ratios1 * np_advs
np_surr2 = np_ratios2 * np_advs
np_surr = np.minimum(np_surr1, np_surr2)
np_loss = - np.sum(np_surr)

print("np_advs:\n", np_advs)
print("np_ratios1:\n", np_ratios1)
print("np_ratios2:\n", np_ratios2)
print("np_surr:\n", np_surr)


# THEANO

th_ratios1 = T.vector('ratios1')
th_ratios2 = T.vector('ratios2')
th_advs = T.vector('th_advs')
th_surr1 = th_ratios1 * th_advs
th_loss1 = - T.sum(th_surr1)
th_grad1 = T.grad(th_loss1, wrt=th_advs)
th_surr2 = th_ratios2 * th_advs
th_loss2 = - T.sum(th_surr2)
th_grad2 = T.grad(th_loss2, wrt=th_advs)
th_surr = T.minimum(th_surr1, th_surr2)
th_loss = - T.sum(th_surr)
th_grad = T.grad(th_loss, wrt=th_advs)

th_f = theano.function(
    inputs=[th_ratios1, th_ratios2, th_advs],
    outputs=[th_surr1, th_surr2, th_surr, th_grad1, th_grad2, th_grad]
)

th_surr1_out, th_surr2_out, th_surr_out, th_grad1_out, th_grad2_out, th_grad_out = \
    th_f(np_ratios1, np_ratios1, np_advs)


print("\nth_surr1:\n", th_surr1_out)
print("th_surr2:\n", th_surr2_out)
print("th_surr:\n", th_surr_out)
print("\nth_grad1:\n", th_grad1_out)
print("th_grad2:\n", th_grad2_out)
print("th_grad:\n", th_grad_out)


# Tensorflow
tf_ratios1 = tf.placeholder(dtype=tf.float32, shape=[None])
tf_ratios2 = tf.placeholder(dtype=tf.float32, shape=[None])
tf_advs = tf.placeholder(dtype=tf.float32, shape=[None])
tf_surr1 = tf_ratios1 * tf_advs
tf_loss1 = - tf.reduce_sum(tf_surr1)
tf_grad1 = tf.gradients(tf_loss1, [tf_advs])
tf_surr2 = tf_ratios2 * tf_advs
tf_loss2 = - tf.reduce_sum(tf_surr2)
tf_grad2 = tf.gradients(tf_loss2, [tf_advs])
tf_surr = tf.minimum(tf_surr1, tf_surr2)
tf_loss = - tf.reduce_sum(tf_surr)
tf_grad = tf.gradients(tf_loss, [tf_advs])

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    tf_surr1_out, tf_surr2_out, tf_surr_out, tf_grad1_out, tf_grad2_out, tf_grad_out = \
        sess.run([tf_surr1, tf_surr2, tf_surr, tf_grad1, tf_grad2, tf_grad],
            feed_dict={tf_ratios1: np_ratios1, tf_ratios2: np_ratios1, tf_advs: np_advs})

print("\ntf_surr1:\n", tf_surr1_out)
print("tf_surr2:\n", tf_surr2_out)
print("tf_surr:\n", tf_surr_out)
print("\ntf_grad1:\n", tf_grad1_out)
print("tf_grad2:\n", tf_grad2_out)
print("tf_grad:\n", tf_grad_out)


assert np.allclose(th_surr1_out, tf_surr1_out)
assert np.allclose(th_surr2_out, tf_surr2_out)
assert np.allclose(th_surr_out, tf_surr_out)
assert np.allclose(th_grad1_out, tf_grad1_out)
assert np.allclose(th_grad2_out, tf_grad2_out)
assert np.allclose(th_grad_out, tf_grad_out)

print("\npassed all tests\n")
```

**Comment by astooke** [comment]
Possibly the problem comes from `theano/scalar/basic.py`, in the classes `Maximum(BinaryScalarOp)` and `Minimum(BinaryScalarOp)`.  In both cases, the last 3 lines of the `L_op()` method reads as follows: (see [Maximum here](https://github.com/Theano/Theano/blob/a44a3f8f1315cf85899a6c17f2916e443f16ef7c/theano/scalar/basic.py#L1654-L1656) and [Minimum here](https://github.com/Theano/Theano/blob/a44a3f8f1315cf85899a6c17f2916e443f16ef7c/theano/scalar/basic.py#L1689-L1691))

```
gx = eq(outputs[0], x) * gz
gy = eq(outputs[0], y) * gz
return (gx, gy)
```
In the above example, both `gx` and `gy` come out non-zero, but they both refer to gradients wrt the same input, so the total amount is apparently getting summed elsewhere.  I think the standard way to handle this non-differentiability is to enforce one of the derivatives to zero, such as:

```
if eq(outputs[0], x):
    gx = gz 
    gy = 0
elif eq(outputs[0], y):
    gx = 0
    gy = gz
else:
    raise ValueError
return (gx, gy)
```

Or possibly could define is to take one-half value at equality:
```
gx = eq(outputs[0], x) * gz
gy = eq(outputs[0], y) * gz
if gx and gy:
    gx *= 0.5
    gy *= 0.5
return (gx, gy)
```

p.s. How I found this was through the use of `clip()`.  The two arguments I passed to m`minimum` were a linear function of a variable (e.g. `a * r`), and the same linear function of a clipped version of the variable (e.g. `a * clipped_r`).   The gradient was with respect to the coefficient `a`.   Whenever the variable was _not_ clipped, the clipped and un-clipped values were equal, so it was counting the gradients from both and summing them.

**Comment by astooke** [comment]
seems to maintain the right types, this works:

```
gx = eq(outputs[0], x) * gz
gy = (1 - eq(outputs[0], x) * gz
return (gx, gy)
```

**Comment by nouiz** [comment]
Do you want to do it? That look pretty good (your last comment). @lamblin also agreed with that fix. If this can be merged by thursday, it would be good as we would like to make a new beta this week.

**Comment by astooke** [comment]
Great!  It's so small, and I don't have the PR workflow in place...happy to have you include it in another PR.

The fix needs to go in `Minimum` and `Maximum`

**Comment by notoraptor** [comment]
Hi here. I think I am experiencing similar issues related to gradient of maximum for CPU spatial transformation implementation ( #6298 ).

I try to apply the last suggested fix, but it does not work in all cases:
```
$ git diff
```
```patch
diff --git a/theano/gradient.py b/theano/gradient.py
index c0cedd4e5..fe0fce47a 100644
--- a/theano/gradient.py
+++ b/theano/gradient.py
@@ -1490,6 +1490,7 @@ class numeric_grad(object):
 
         if eps is None:
             eps = builtins.max(self.type_eps[dt] for dt in dtypes)
+        self.eps = eps
 
         # set up aliases so that apt[i] is backed by memory in x
         # and self.gf is backed by memory in gx
@@ -1786,6 +1787,12 @@ def verify_grad(fun, pt, n_tests=2, rng=None, eps=None,
                 analytic_grad, abs_tol, rel_tol)
 
             if max_abs_err > abs_tol and max_rel_err > rel_tol:
+                print('Failing gradient case (argument %s/%s):' % (max_arg + 1, len(analytic_grad)))
+                print('Eps:', num_grad.eps)
+                print('Expected:', num_grad.gf[max_arg].dtype, num_grad.gf[max_arg].shape)
+                print(num_grad.gf[max_arg])
+                print('Value:', analytic_grad[max_arg].dtype, analytic_grad[max_arg].shape)
+                print(analytic_grad[max_arg])
 
                 raise verify_grad.E_grad(max_arg, max_err_pos,
                                          analytic_grad[max_arg].shape,
diff --git a/theano/scalar/basic.py b/theano/scalar/basic.py
index fe2dbe735..6ef59d1bc 100644
--- a/theano/scalar/basic.py
+++ b/theano/scalar/basic.py
@@ -1652,7 +1652,7 @@ class Maximum(BinaryScalarOp):
                     y.zeros_like().astype(theano.config.floatX)]
 
         gx = eq(outputs[0], x) * gz
-        gy = eq(outputs[0], y) * gz
+        gy = (1 - eq(outputs[0], x)) * gz
         return (gx, gy)
 
 maximum = Maximum(upcast_out, name='maximum')
```

Then I test with following script:
```python
from __future__ import print_function
import numpy as np
import theano
from theano.tests import unittest_tools as utt

def functor_x_x(x):
	return theano.tensor.maximum(x, x)

def functor_x_2x(x):
	return theano.tensor.maximum(x, x * 2)

def functor_2x_x(x):
	return theano.tensor.maximum(x * 2, x)

x_val = np.arange(20, dtype=theano.config.floatX)
# x_val[0] will be always 0 for x_val or x_val * 2.

print('Verifying gradient of max(x, x)')
utt.verify_grad(functor_x_x, [x_val], n_tests=10)

print('Verifying gradient of max(2x, x)')
utt.verify_grad(functor_2x_x, [x_val], n_tests=10)

print('Verifying gradient of max(x, 2x)')
utt.verify_grad(functor_x_2x, [x_val], n_tests=10)

print('OK')
```

And this is the output:
```
Verifying gradient of max(x, x)
Verifying gradient of max(2x, x)
Verifying gradient of max(x, 2x)
Failing gradient case (argument 1/1):
Eps: 1e-07
Expected: float64 (20,)
[ 2.40087388  2.68837312  2.35302878  2.45571584  2.90291553  1.02540639
  1.82717542  1.09762539  1.19985657  2.01613261  1.40049508  2.48830816
  1.385784    2.40168958  1.58645662  2.54895895  1.01021783  1.22571521
  1.22190727  1.49533662]
Value: float64 (20,)
[ 1.20043712  2.68837329  2.35302867  2.45571611  2.90291591  1.02540639
  1.8271754   1.09762559  1.19985712  2.01613261  1.40049508  2.48830834
  1.38578401  2.4016895   1.58645621  2.54895891  1.01021768  1.22571531
  1.22190734  1.49533646]
Traceback (most recent call last):
  File "../utils/python/tn.py", line 25, in <module>
    utt.verify_grad(functor_x_2x, [x_val], n_tests=10)
  File "/mnt/00447B2C447B2396/Users/notoraptor/mila/dev/git/theano/theano/tests/unittest_tools.py", line 92, in verify_grad
    T.verify_grad(op, pt, n_tests, rng, *args, **kwargs)
  File "/mnt/00447B2C447B2396/Users/notoraptor/mila/dev/git/theano/theano/gradient.py", line 1802, in verify_grad
    abs_tol, rel_tol)
theano.gradient.GradientError: GradientError: numeric gradient and analytic gradient exceed tolerance:
        At position 0 of argument 0 with shape (20,),
            val1 = 1.200437      ,  val2 = 2.400874
            abs. error = 1.200437,  abs. tolerance = 0.000100
            rel. error = 0.333333,  rel. tolerance = 0.000100
Exception args: 
The error happened with the following inputs:, [array([  0.,   1.,   2.,   3.,   4.,   5.,   6.,   7.,   8.,   9.,  10.,
        11.,  12.,  13.,  14.,  15.,  16.,  17.,  18.,  19.])], 
The value of eps is:, None, 
The out_type is:, None
```

So, gradient now works for `max(x, x)` and `max(2x, x)` but fails for `max(x, 2x)` because gradient is not computed well for `x[0]`. `x[0]` is `0`, so `2 * x[0]` is still `0`, so gradient for `max(x[0], 2*x[0])` is a case when both arguments are equal. It seems current fix works only when the left argument is modified.

In fact, I think `max(x, y)` is not differentiable when we rigorously have `x == y`.

@nouiz @lamblin

**Comment by nouiz** [comment]
I just did a PR for this issue here: https://github.com/Theano/Theano/pull/6477

@notoraptor to test my fix, I was not able to test it with verify grad, as it is at a discontinuity point. I tested the graph manually with the expected output at that point.

At discontinuity point, we normally use the value just before or after the discontinuity as for absolute. My PR continue in that direction, but make us not able to use verify_grad at that point.
